prov/gni/src/gnix_rma.c

```diff
@@ -995,6 +1185,42 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
 		req->flags |= GNIX_RMA_RDMA;
 	}
 
+	/* Add reads/writes to slist when FI_MORE is present, Or
+	 * if this is the first message in the chain without FI_MORE.
+	 * When FI_MORE is not present, if the slists are not empty
+	 * it is the first message without FI_MORE. */
+	if ((flags & FI_MORE) ||
+	    (!(slist_empty(&ep->more_write)) || !(slist_empty(&ep->more_read)))) {
+		req->work_fn = _gnix_rma_more_post_req;
+		if (fr_type == GNIX_FAB_RQ_RDMA_READ) {
+			slist_insert_tail(&req->rma.sle, &ep->more_read);
+		}
+		else {
+			slist_insert_tail(&req->rma.sle, &ep->more_write);
+		}
+		if (flags & FI_MORE)
+			return FI_SUCCESS;
+	}
+
+	/* Initiate reads and writes on first message without FI_MORE */
+	if (!(flags & FI_MORE) &&
+	    (!(slist_empty(&ep->more_write)) || !(slist_empty(&ep->more_read)))) {
+		if (!(slist_empty(&ep->more_write))) {
+			sle = ep->more_write.head;
+			more_req = container_of(sle, struct gnix_fab_req, rma.sle);
+			GNIX_INFO(FI_LOG_EP_DATA, "MORE: got fab_request from more_write. Queuing Request\n");
+			_gnix_vc_queue_tx_req(more_req);
+			slist_init(&ep->more_write); /* Empty slist for future fab_reqs */
+		}
+		if (!(slist_empty(&ep->more_read))) {
+			sle = ep->more_read.head;
+			more_req = container_of(sle, struct gnix_fab_req, rma.sle);
```

Comment:
slist_first_entry() is a glaring omission from our list API.  It would be nice to add this to prov/gni/include/gnix_util.h to clean up these lines.